File: thought-vault/app/src/main/java/com/example/thoughtvault/data/remote/WebdavClient.kt
package com.example.thoughtvault.data.remote

import android.util.Base64
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import timber.log.Timber
import java.io.IOException
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.nio.charset.StandardCharsets
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * WebDAV 客户端 — 所有 NAS 通信的唯一入口。
 * 封装了 HTTP Basic Auth、PUT/GET/MKCOL/PROPFIND/DELETE 操作。
 */
@Singleton
class WebdavClient @Inject constructor() {

    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)
            .addInterceptor { chain ->
                val request = chain.request()
                val response = chain.proceed(request)
                Timber.d("${request.method} ${request.url} -> ${response.code}")
                response
            }
            .build()
    }

    /**
     * 指数退避重试：遇到网络瞬时错误时自动重试。
     * 延迟序列：500ms → 1s → 2s（最多 3 次尝试）
     */
    private suspend fun <T> retryWithBackoff(
        maxAttempts: Int = 3,
        initialDelayMs: Long = 500,
        block: suspend () -> Result<T>,
    ): Result<T> {
        var attempt = 0
        var lastResult: Result<T>? = null

        while (attempt < maxAttempts) {
            val result = block()
            if (result.isSuccess) return result

            val error = result.exceptionOrNull()
            // 仅对可重试的错误（网络瞬时故障）进行重试
            if (error is ConnectException || error is SocketTimeoutException) {
                attempt++
                if (attempt < maxAttempts) {
                    val delay = initialDelayMs * (1L shl (attempt - 1))  // 500, 1000, 2000
                    Timber.d("重试 $attempt/$maxAttempts，等待 ${delay}ms...")
                    kotlinx.coroutines.delay(delay)
                    lastResult = result
                    continue
                }
            }
            // 不可重试的错误（4xx, 5xx），直接返回
            return result
        }
        return lastResult ?: Result.failure(IOException("重试耗尽"))
    }

    /** 构建带 Basic Auth 的请求头 */
    private fun authHeaders(username: String, password: String): Map<String, String> {
        val credentials = "$username:$password"
        val encoded = Base64.encodeToString(
            credentials.toByteArray(StandardCharsets.UTF_8), Base64.NO_WRAP
        )
        return mapOf("Authorization" to "Basic $encoded")
    }

    /**
     * PUT — 创建或覆盖文件（含指数退避重试）。
     */
    suspend fun put(
        url: String,
        content: String,
        username: String,
        password: String,
    ): Result<Unit> = withContext(Dispatchers.IO) {
        retryWithBackoff {
            try {
                val body = content.toRequestBody("text/markdown; charset=utf-8".toMediaType())
                val request = Request.Builder()
                    .url(url)
                    .apply { authHeaders(username, password).forEach { (k, v) -> addHeader(k, v) } }
                    .put(body)
                    .build()

                val response = client.newCall(request).execute()
                if (response.isSuccessful) {
                    Result.success(Unit)
                } else {
                    Result.failure(IOException("PUT 失败: HTTP ${response.code}"))
                }
            } catch (e: ConnectException) {
                Result.failure(ConnectionException("无法连接到 NAS", e))
            } catch (e: SocketTimeoutException) {
                Result.failure(ConnectionException("连接超时", e))
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * GET — 读取文件内容。
     * 返回 null 表示文件不存在（404）。
     */
    suspend fun get(
        url: String,
        username: String,
        password: String,
    ): Result<String?> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url(url)
                .apply { authHeaders(username, password).forEach { (k, v) -> addHeader(k, v) } }
                .get()
                .build()

            val response = client.newCall(request).execute()
            when {
                response.isSuccessful -> Result.success(response.body?.string() ?: "")
                response.code == 404 -> Result.success(null)
                else -> Result.failure(IOException("GET failed: ${response.code}"))
            }
        } catch (e: ConnectException) {
            Result.failure(ConnectionException("无法连接到 NAS", e))
        } catch (e: SocketTimeoutException) {
            Result.failure(ConnectionException("连接超时", e))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * PUT — 创建空文件（用于锁文件）。
     */
    suspend fun putEmpty(
        url: String,
        username: String,
        password: String,
    ): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url(url)
                .apply { authHeaders(username, password).forEach { (k, v) -> addHeader(k, v) } }
                .put("".toRequestBody(null))
                .build()

            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                Result.success(Unit)
            } else {
                Result.failure(IOException("PUT empty failed: ${response.code}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * DELETE — 删除文件（用于释放锁）。
     */
    suspend fun delete(
        url: String,
        username: String,
        password: String,
    ): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url(url)
                .apply { authHeaders(username, password).forEach { (k, v) -> addHeader(k, v) } }
                .delete()
                .build()

            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                Result.success(Unit)
            } else {
                Result.failure(IOException("DELETE failed: ${response.code}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * MKCOL — 创建目录（按需创建年/月子目录）。
     */
    suspend fun mkcol(
        url: String,
        username: String,
        password: String,
    ): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url(url)
                .apply { authHeaders(username, password).forEach { (k, v) -> addHeader(k, v) } }
                .method("MKCOL", null)
                .build()

            val response = client.newCall(request).execute()
            // 201 Created 或 405 Method Not Allowed（目录已存在）都视为成功
            if (response.isSuccessful || response.code == 405) {
                Result.success(Unit)
            } else {
                Result.failure(IOException("MKCOL failed: ${response.code}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * PROPFIND — 列出目录内容（用于验证连接和浏览历史）。
     */
    suspend fun propfind(
        url: String,
        username: String,
        password: String,
        depth: Int = 1,
    ): Result<String> = withContext(Dispatchers.IO) {
        try {
            val body = """<?xml version="1.0" encoding="utf-8"?>
                <D:propfind xmlns:D="DAV:">
                  <D:prop>
                    <D:displayname/>
                    <D:getlastmodified/>
                    <D:getcontentlength/>
                  </D:prop>
                </D:propfind>""".trimIndent()
                .toRequestBody("application/xml; charset=utf-8".toMediaType())

            val request = Request.Builder()
                .url(url)
                .apply { authHeaders(username, password).forEach { (k, v) -> addHeader(k, v) } }
                .header("Depth", depth.toString())
                .method("PROPFIND", body)
                .build()

            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                Result.success(response.body?.string() ?: "")
            } else {
                Result.failure(IOException("PROPFIND failed: ${response.code}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * 测试 NAS 连接是否正常。
     * 向根目录发送 PROPFIND Depth:0 请求。
     */
    suspend fun testConnection(baseUrl: String, username: String, password: String): Result<Boolean> {
        return propfind(baseUrl, username, password, depth = 0).map { true }
    }
}

/** 连接异常 */
class ConnectionException(message: String, cause: Throwable? = null) : IOException(message, cause)

File: sanitized/thought-vault-web/src/lib/webdav-client.ts
import { type Result, ok, err } from './types'

/**
 * WebDAV 底层客户端 — 对标 Android WebdavClient.kt
 * 所有请求通过同域 /api/ 转发到 NAS Python 代理
 */
export class WebdavClient {
  private baseUrl: string
  private username: string
  private password: string
  private proxyPath: string

  constructor(baseUrl: string, username: string, password: string, proxyPath = '/api/') {
    this.baseUrl = baseUrl.replace(/\/+$/, '')
    this.username = username
    this.password = password
    this.proxyPath = proxyPath
  }

  /** 构建代理 URL — 路径直接透传，不再使用 ?path= 查询参数 */
  private proxyUrl(webdavPath: string): string {
    const path = webdavPath.replace(/^\//, '')
    return `${this.proxyPath}${path}`
  }

  /** HTTP Basic Auth 头 */
  private authHeaders(): Record<string, string> {
    const creds = btoa(`${this.username}:${this.password}`)
    return {
      'Authorization': `Basic ${creds}`,
      'Content-Type': 'text/markdown; charset=utf-8',
    }
  }

  /** 指数退避重试（对标 Android retryWithBackoff） */
  private async retryWithBackoff<T>(
    fn: () => Promise<Result<T>>,
    maxAttempts = 3,
    initialDelayMs = 500,
  ): Promise<Result<T>> {
    let lastError: Result<T> | null = null
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      const result = await fn()
      if (result.success) return result
      lastError = result

      // 仅对网络错误重试
      if (attempt < maxAttempts - 1) {
        const delay = initialDelayMs * (1 << attempt) // 500, 1000, 2000
        console.log(`重试 ${attempt + 1}/${maxAttempts}，等待 ${delay}ms...`)
        await new Promise(r => setTimeout(r, delay))
      }
    }
    return lastError!
  }

  /** PUT — 创建或覆盖文件 */
  async put(webdavPath: string, content: string): Promise<Result<void>> {
    return this.retryWithBackoff(async () => {
      try {
        const resp = await fetch(this.proxyUrl(webdavPath), {
          method: 'PUT',
          headers: this.authHeaders(),
          body: content,
        })
        if (resp.ok) return ok(undefined)
        return err(`PUT 失败: HTTP ${resp.status}`)
      } catch (e) {
        return err(e instanceof Error ? e.message : 'PUT 网络错误')
      }
    })
  }

  /** GET — 读取文件内容，404 返回 null */
  async get(webdavPath: string): Promise<Result<string | null>> {
    try {
      const resp = await fetch(this.proxyUrl(webdavPath), {
        headers: { 'Authorization': this.authHeaders()['Authorization'] },
      })
      if (resp.ok) {
        const text = await resp.text()
        return ok(text)
      }
      if (resp.status === 404) return ok(null)
      return err(`GET 失败: HTTP ${resp.status}`)
    } catch (e) {
      return err(e instanceof Error ? e.message : 'GET 网络错误')
    }
  }

  /** PUT 空内容（锁文件） */
  async putEmpty(webdavPath: string): Promise<Result<void>> {
    try {
      const resp = await fetch(this.proxyUrl(webdavPath), {
        method: 'PUT',
        headers: { 'Authorization': this.authHeaders()['Authorization'] },
      })
      if (resp.ok) return ok(undefined)
      return err(`PUT empty 失败: HTTP ${resp.status}`)
    } catch (e) {
      return err(e instanceof Error ? e.message : 'PUT empty 网络错误')
    }
  }

  /** DELETE */
  async delete(webdavPath: string): Promise<Result<void>> {
    try {
      const resp = await fetch(this.proxyUrl(webdavPath), {
        method: 'DELETE',
        headers: { 'Authorization': this.authHeaders()['Authorization'] },
      })
      if (resp.ok) return ok(undefined)
      return err(`DELETE 失败: HTTP ${resp.status}`)
    } catch (e) {
      return err(e instanceof Error ? e.message : 'DELETE 网络错误')
    }
  }

  /** MKCOL — 创建目录 */
  async mkcol(webdavPath: string): Promise<Result<void>> {
    try {
      const resp = await fetch(this.proxyUrl(webdavPath), {
        method: 'POST',
        headers: {
          ...this.authHeaders(),
          'X-Http-Method-Override': 'MKCOL',
        },
      })
      if (resp.ok || resp.status === 405) return ok(undefined)
      return err(`MKCOL 失败: HTTP ${resp.status}`)
    } catch (e) {
      return err(e instanceof Error ? e.message : 'MKCOL 网络错误')
    }
  }

  /** PROPFIND — 列出目录内容。depth: 0 | 1 | "infinity"（Synology Apache 不接受 2） */
  async propfind(webdavPath: string, depth: number | string = 1): Promise<Result<string>> {
    const body = `<?xml version="1.0" encoding="utf-8"?>
<D:propfind xmlns:D="DAV:">
  <D:prop>
    <D:displayname/>
    <D:getlastmodified/>
    <D:getcontentlength/>
  </D:prop>
</D:propfind>`.trim()

    // 通过 nginx /api/ 代理访问 WebDAV，复用容器的 LE 证书
    try {
      const resp = await fetch(this.proxyUrl(webdavPath), {
        method: 'PROPFIND',
        headers: {
          ...this.authHeaders(),
          'Content-Type': 'application/xml; charset=utf-8',
          'Depth': String(depth),
        },
        body,
      })
      if (resp.ok) {
        const text = await resp.text()
        return ok(text)
      }
      return err(`PROPFIND 失败: HTTP ${resp.status}`)
    } catch (e) {
      return err(e instanceof Error ? e.message : 'PROPFIND 网络错误')
    }
  }

  /** 测试连接 */
  async testConnection(): Promise<Result<boolean>> {
    const result = await this.propfind('', 0)
    if (result.success) return ok(true)
    return err(result.error || '连接测试失败')
  }
}
